core/imt/inc/ROOT/TExecutor.hxx

```diff
@@ -0,0 +1,494 @@
+template<class F, class INTEGER, class R, class Cond = noReferenceCond<F, INTEGER>>
+auto MapReduce(F func, ROOT::TSeq<INTEGER> args, R redfunc, unsigned nChunks) -> typename std::result_of<F(INTEGER)>::type;
+template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+auto MapReduce(F func, std::initializer_list<T> args, R redfunc, unsigned nChunks) -> typename std::result_of<F(T)>::type;
+template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+auto MapReduce(F func, std::vector<T> &args, R redfunc, unsigned nChunks) -> typename std::result_of<F(T)>::type;
+template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+auto MapReduce(F func, const std::vector<T> &args, R redfunc, unsigned nChunks) -> typename std::result_of<F(T)>::type;
+
+using TExecutorCRTP<TExecutor>::Reduce;
+
+unsigned GetPoolSize();
+
+protected:
+template<class F, class R, class Cond = noReferenceCond<F>>
+auto Map(F func, unsigned nTimes, R redfunc, unsigned nChunks) -> std::vector<typename std::result_of<F()>::type>;
+template<class F, class INTEGER, class R, class Cond = noReferenceCond<F, INTEGER>>
+auto Map(F func, ROOT::TSeq<INTEGER> args, R redfunc, unsigned nChunks) -> std::vector<typename std::result_of<F(INTEGER)>::type>;
+template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+auto Map(F func, std::vector<T> &args, R redfunc, unsigned nChunks) -> std::vector<typename std::result_of<F(T)>::type>;
+template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+auto Map(F func, const std::vector<T> &args, R redfunc, unsigned nChunks) -> std::vector<typename std::result_of<F(T)>::type>;
+template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+auto Map(F func, std::initializer_list<T> args, R redfunc, unsigned nChunks) -> std::vector<typename std::result_of<F(T)>::type>;
+
+private:
+   ROOT::Internal::ExecutionPolicy fExecPolicy;
+#ifdef R__USE_IMT
+   std::unique_ptr<ROOT::TThreadExecutor> fThreadPool;
+#endif
+#ifndef R__WIN32
+   std::unique_ptr<ROOT::TProcessExecutor> fProcPool;
+#endif
+   std::unique_ptr<ROOT::TSequentialExecutor> fSeqPool;
+};
+
+
+//////////////////////////////////////////////////////////////////////////
+/// \copydoc TExecutorCRTP::Map(F func,unsigned nTimes)
+template<class F, class Cond>
+auto TExecutor::Map(F func, unsigned nTimes) -> std::vector<typename std::result_of<F()>::type> {
+   using retType = decltype(func());
+   std::vector<retType> res;
+   switch(fExecPolicy){
+      case ROOT::Internal::ExecutionPolicy::kSerial:
+         res = fSeqPool->Map(func, nTimes);
+         break;
+#ifdef R__USE_IMT
+      case ROOT::Internal::ExecutionPolicy::kMultithread:
+         res = fThreadPool->Map(func, nTimes);
+         break;
+#endif
+#ifndef R__WIN32
+      case ROOT::Internal::ExecutionPolicy::kMultiprocess:
+         res = fProcPool->Map(func, nTimes);
+         break;
+#endif
+      default:
+         break;
+   }
```

Comment:
This needs refactoring / reduction in code dupe - @hageboeck might have an idea that he'll share :-)

Reply:
I could get rid of all the defines, but I can't refactor the switch statement. I would need to figure out the return type from the Map arguments, and this is not possible to do in a general way.